pkgs/java_http/lib/src/java_client.dart

```diff
@@ -39,51 +41,103 @@ class JavaClient extends BaseClient {
-
-      httpUrlConnection.disconnect();
-
-      return (
-        statusCode,
-        reasonPhrase,
-        responseHeaders,
-        responseBody,
-      );
-    });
+    final isolateRequest = (
+      sendPort: receivePort.sendPort,
+      url: request.url,
+      method: request.method,
+      headers: request.headers,
+      body: await request.finalize().toBytes(),
+    );
+
+    // Could create a new class to hold the data for the isolate instead
+    // of using a record.
+    await Isolate.spawn(_isolateMethod, isolateRequest);
+
+    final statusCode = await events.next as int;
+    final reasonPhrase = await events.next as String?;
+    final responseHeaders = await events.next as Map<String, String>;
+
+    Stream<List<int>> responseBodyStream(Stream<dynamic> events) async* {
+      try {
+        await for (final event in events) {
+          if (event is List<int>) {
+            yield event;
+          } else if (event is ClientException) {
+            throw event;
+          } else if (event == null) {
+            return;
+          }
+        }
+      } finally {
+        // TODO: Should we kill the isolate here?
+        receivePort.close();
+      }
+    }
 
-    return StreamedResponse(Stream.value(responseBody), statusCode,
-        contentLength:
-            _contentLengthHeader(request, responseHeaders, responseBody.length),
+    return StreamedResponse(responseBodyStream(events.rest), statusCode,
+        contentLength: _parseContentLengthHeader(request.url, responseHeaders),
         request: request,
         headers: responseHeaders,
         reasonPhrase: reasonPhrase);
   }
 
+  // TODO: Rename _isolateMethod to something more descriptive.
+  void _isolateMethod(
+    ({
+      SendPort sendPort,
+      Uint8List body,
+      Map<String, String> headers,
+      String method,
+      Uri url,
+    }) request,
+  ) {
+    final httpUrlConnection = URL
+        .ctor3(request.url.toString().toJString())
+        .openConnection()
+        .castTo(HttpURLConnection.type, deleteOriginal: true);
+
+    request.headers.forEach((headerName, headerValue) {
+      httpUrlConnection.setRequestProperty(
+          headerName.toJString(), headerValue.toJString());
+    });
+
+    httpUrlConnection.setRequestMethod(request.method.toJString());
+    _setRequestBody(httpUrlConnection, request.body);
+
+    final statusCode = _statusCode(request.url, httpUrlConnection);
+    request.sendPort.send(statusCode);
+
+    final reasonPhrase = _reasonPhrase(httpUrlConnection);
+    request.sendPort.send(reasonPhrase);
+
+    final responseHeaders = _responseHeaders(httpUrlConnection);
+    request.sendPort.send(responseHeaders);
+
+    // TODO: Throws a ClientException if the content length header is invalid.
+    // I think we need to send the ClientException over the SendPort.
+    final contentLengthHeader = _parseContentLengthHeader(
+      request.url,
+      responseHeaders,
+    );
```

Comment:
I think this is where the unhandled ClientException is being thrown inside the worker isolate.

Reply:
I'll make this change in a separate PR.